cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -313,78 +233,67 @@ func deploymentNameForEnv(envName string, clock clock.Clock) string {
 }
 
 // Provisioning the infrastructure within the specified template
-func (p *BicepProvider) Deploy(
-	ctx context.Context,
-	pd *DeploymentPlan,
-) *async.InteractiveTaskWithProgress[*DeployResult, *DeployProgress] {
-	return async.RunInteractiveTaskWithProgress(
-		func(asyncContext *async.InteractiveTaskContextWithProgress[*DeployResult, *DeployProgress]) {
-			done := make(chan bool)
-
-			// Ensure the done marker channel is sent in all conditions
-			defer func() {
-				done <- true
-			}()
-
-			bicepDeploymentData := pd.Details.(BicepDeploymentDetails)
-
-			// Report incremental progress
-			go func() {
-				resourceManager := infra.NewAzureResourceManager(p.azCli)
-				progressDisplay := NewProvisioningProgressDisplay(resourceManager, p.console, bicepDeploymentData.Target)
-				// Make initial delay shorter to be more responsive in displaying initial progress
-				initialDelay := 3 * time.Second
-				regularDelay := 10 * time.Second
-				timer := time.NewTimer(initialDelay)
-				queryStartTime := time.Now()
-
-				for {
-					select {
-					case <-done:
-						timer.Stop()
-						return
-					case <-timer.C:
-						if progressReport, err := progressDisplay.ReportProgress(ctx, &queryStartTime); err == nil {
-							asyncContext.SetProgress(progressReport)
-						} else {
-							// We don't want to fail the whole deployment if a progress reporting error occurs
-							log.Printf("error while reporting progress: %s", err.Error())
-						}
-
-						timer.Reset(regularDelay)
-					}
-				}
-			}()
-
-			// Start the deployment
-			p.console.ShowSpinner(ctx, "Creating/Updating resources", input.Step)
-
-			deployResult, err := p.deployModule(
-				ctx,
-				bicepDeploymentData.Target,
-				bicepDeploymentData.Template,
-				bicepDeploymentData.Parameters,
-				map[string]*string{
-					azure.TagKeyAzdEnvName: to.Ptr(p.env.GetEnvName()),
-				},
-			)
-			if err != nil {
-				asyncContext.SetError(err)
+func (p *BicepProvider) Deploy(ctx context.Context, pd *DeploymentPlan) (*DeployResult, error) {
+	done := make(chan bool)
+
+	// Ensure the done marker channel is sent in all conditions
+	defer func() {
+		done <- true
+	}()
+
+	bicepDeploymentData := pd.Details.(BicepDeploymentDetails)
+
+	// Report incremental progress
+	go func() {
```

Comment:
bicep_provider can avoid having this go routine.
Here's why:
the call to `p.deployModule()` eventually calls `deploymentClient.BeginCreateOrUpdateAtSubscriptionScope()`.  Depending on scope, it could be subscription or resource group, but using the same client `deploymentClient`.
Then it uses `PollUntilDone()`, which is the reason why bicep_provider is requiring a go routine for showing the progress while waiting for `PollUntilDone()`.

However, the bicep_provider could instead receive the `poller` back and display the progress while the poller is not done.  So, instead of making the go routine to check the done-channel to exit, we would have a simple for loop checking the poller status before running the progress loop again.

to make this change, we need to change the scope.Deploy() to return a poller[T] instead of waiting for the T.

I will use this comment to make a new issue for this, as it would make this PR bigger

Reply:
https://github.com/Azure/azure-dev/issues/2338

Reply:
Thanks, I'm going to leave this as a simple result right now and we can discuss this with the overall updates to how we report progress back to the commands.